From 9f26eeb5a71cf418d318a5866cfdc49e98d737e3 Mon Sep 17 00:00:00 2001
From: AlOa <orlandi.alberto@gmail.com>
Date: Wed, 15 Nov 2023 10:08:31 +0100
Subject: [PATCH 17/54] Prelu layer uses  sse instruction _mm_load_ps but data
 can be misaligned so it must use _mm_loadu_ps (#5149)

---
 src/layer/x86/prelu_x86.cpp | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/layer/x86/prelu_x86.cpp b/src/layer/x86/prelu_x86.cpp
index 52334659..95c0560c 100644
--- a/src/layer/x86/prelu_x86.cpp
+++ b/src/layer/x86/prelu_x86.cpp
@@ -81,7 +81,7 @@ int PReLU_x86::forward_inplace(Mat& bottom_top_blob, const Option& opt) const
             {
                 int i = remain_size_start + ii * 4;
                 __m128 _p128 = _mm_load_ps(ptr + i);
-                __m128 _slope128 = _mm_load_ps(slope + i);
+                __m128 _slope128 = _mm_loadu_ps(slope + i);
                 _mm_store_ps(ptr + i, prelu_sse(_p128, _slope128));
             }
             remain_size_start += nn_size * 4;
@@ -157,7 +157,7 @@ int PReLU_x86::forward_inplace(Mat& bottom_top_blob, const Option& opt) const
 
             float slope = num_slope > 1 ? slope_data[i] : slope_data[0];
 #if __SSE2__
-            __m128 _slope128 = num_slope > 1 && (elempack == 4) ? _mm_load_ps((const float*)slope_data + i * 4) : _mm_set1_ps(slope);
+            __m128 _slope128 = num_slope > 1 && (elempack == 4) ? _mm_loadu_ps((const float*)slope_data + i * 4) : _mm_set1_ps(slope);
 #if __AVX__
             __m256 _slope256 = num_slope > 1 && (elempack == 8) ? _mm256_loadu_ps((const float*)slope_data + i * 8) : _mm256_insertf128_ps(_mm256_castps128_ps256(_slope128), _slope128, 1);
 #if __AVX512F__
@@ -205,7 +205,7 @@ int PReLU_x86::forward_inplace(Mat& bottom_top_blob, const Option& opt) const
 
             float slope = num_slope > 1 ? slope_data[q] : slope_data[0];
 #if __SSE2__
-            __m128 _slope128 = num_slope > 1 && (elempack == 4) ? _mm_load_ps((const float*)slope_data + q * 4) : _mm_set1_ps(slope);
+            __m128 _slope128 = num_slope > 1 && (elempack == 4) ? _mm_loadu_ps((const float*)slope_data + q * 4) : _mm_set1_ps(slope);
 #if __AVX__
             __m256 _slope256 = num_slope > 1 && (elempack == 8) ? _mm256_loadu_ps((const float*)slope_data + q * 8) : _mm256_insertf128_ps(_mm256_castps128_ps256(_slope128), _slope128, 1);
 #if __AVX512F__
-- 
2.47.1

